collections/_article/rendering-acyclic-graph.md

```diff
@@ -0,0 +1,277 @@
+
+### Graph visualization
+
+While this wouldn't provide a direct benefit to end users, building a visualizer for the graph can help the Godot developers have a clear overview of the rendering pipeline of a given frame and identify bottlenecks more easily. During development, a few compute passes were identified that weren't being parallelized correctly due to implementation errors. For example, GPU particle systems were binding an unused buffer for write operations even if they never wrote to it, which led to the commands being identified as being dependent of each other due to having to synchronize with the potential "write" performed by the previous system.
+
+| ![](/storage/blog/acyclic-graph/particle-error.webp) |
+|:--:| 
+| *Due to an implementation error, even after implementing the graph, the execution of the particle systems was mostly linear, as they all reused the same temporary buffer for reading and writing.* |
+
+| ![](/storage/blog/acyclic-graph/particle-error-fix.webp) |
+|:--:| 
+| *After fixing the error by assigning each system their own buffers, the graph automatically reordered and executed the particle systems in parallel, leading to huge gains in performance.* |
+
+While a more obvious case like this one was identified since it did not meet expectations at first, there could be more subtle instances of this behavior yet to be found in the codebase that could be easily exposed by building better debugging tools.
+
+## Conclusions
+
+While the use of a directed acyclic graph for rendering is not a brand new technique, the approach used by Godot is quite novel in many ways. The simplicity of the API exposed by Rendering Device no longer comes at the cost of rendering performance: the developer is guaranteed they'll get efficient use of the GPU and it can only get better from here. As the engine aims to be as general purpose as possible and maintain its ease of use, a lot of alternatives had to be discarded until the right approach was found. This is a long term technical investment that will pay off by reducing the cost of maintenance and unlocking new strategies for optimization in the long run. 
+
+An automatic approach ensures that despite the project being open-source and modified by many different people, they no longer need complete awareness of how the entire rendering pipeline works to modify it. This will also be very beneficial when PRs like [rendering effects](https://github.com/godotengine/godot/pull/80214) are merged, which introduces the ability to add post-processing steps where the Godot renderer will be completely unaware of all dependencies the hook may require. The extra validation introduced by recording commands to the graph has also exposed existing implementation errors, resulting in either bug fixes or even performance improvements.
```

Comment:
```suggestion
An automatic approach ensures that despite the project being open-source and modified by many different people, they no longer need complete awareness of how the entire rendering pipeline works to modify it. This will also be very beneficial when PRs like [rendering effects](https://github.com/godotengine/godot/pull/80214) are merged, which introduce the ability to add post-processing steps where the Godot renderer will be completely unaware of all dependencies the hook may require. The extra validation introduced by recording commands to the graph has also exposed existing implementation errors, resulting in either bug fixes or even performance improvements.
```
Also, I'm not sure what is meant by "An automatic approach".

Reply:
By "an automatic approach" I want to specify that no manual work is required from the engine developers to make the acyclic graph work. It's important to note this as other engines require describing it manually.